rust/src/optimize.rs

```diff
@@ -0,0 +1,464 @@
+//! TODO: Optimize
+
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::hash::{Hash, Hasher};
+use std::sync::Arc;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use log::debug;
+use log::error;
+use parquet::arrow::{ArrowReader, ArrowWriter, ParquetFileArrowReader};
+use parquet::basic::Compression;
+use parquet::file::properties::WriterProperties;
+use parquet::file::serialized_reader::{SerializedFileReader, SliceableCursor};
+use parquet::file::writer::InMemoryWriteableCursor;
+
+use crate::action::{self, Action};
+use crate::{
+    generate_parquet_filename, DeltaDataTypeLong, DeltaTable, DeltaTableError, DeltaTablePartition,
+    PartitionFilter, SchemaTypeStruct,
+};
+
+#[derive(Default, Debug)]
+///Metrics from Optimize
+pub struct Metrics {
+    ///Number of optimized files added
+    pub num_files_added: u64,
+    ///Number of unoptimized files removed
+    pub num_files_removed: u64,
+    ///Detailed metrics for the add operation
+    pub files_added: MetricDetails,
+    ///Detailed metrics for the remove operation
+    pub files_removed: MetricDetails,
+    ///Number of partitions that had at least one file optimized
+    pub partitions_optimized: u64,
+    ///TODO
+    pub num_batches: u64,
+    ///How many files were considered during optimization. Not every file considered is optimized
+    pub total_considered_files: usize,
+    ///How many files were considered for optimization but were skipped
+    pub total_files_skipped: usize,
+    ///TODO
+    pub preserve_insertion_order: bool,
+}
+
+#[derive(Debug)]
+///Statistics on files for a particular operation
+/// Operation can be remove or add
+pub struct MetricDetails {
+    ///Maximum file size of a operation
+    pub min: usize,
+    ///Minimum file size of a operation
+    pub max: usize,
+    ///Average file size of a operation
+    pub avg: f64,
+    ///Number of files encountered during operation
+    pub total_files: usize,
+    ///Sum of file sizes of a operation
+    pub total_size: usize,
+}
+
+impl Default for MetricDetails {
+    fn default() -> Self {
+        MetricDetails {
+            min: usize::MAX,
+            max: 0,
+            avg: 0.0,
+            total_files: 0,
+            total_size: 0,
+        }
+    }
+}
+
+///TODO: Optimize
+#[derive(Default)]
+pub struct Optimize<'a> {
+    filters: &'a [PartitionFilter<'a, &'a str>],
```

Comment:
Should we also have `max_file_size` here? In case the user wishes to pass the value in directly?

Reply:
Done. A user provided target size can now be passed in this builder. Otherwise the table config or default is used.